packages/components/src/internal/AppContext.tsx

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2021 LabKey Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import React, { createContext, FC, useContext, useMemo } from 'react';
+
+import { ComponentsAPIWrapper, getDefaultAPIWrapper } from './APIWrapper';
+
+export interface AppContext {
+    api: ComponentsAPIWrapper;
+}
+
+const Context = createContext<AppContext>(undefined);
+
+export interface AppContextProviderProps {
+    initialContext?: AppContext;
+}
+
+export const AppContextProvider: FC<AppContextProviderProps> = ({ children, initialContext }) => {
+    // Provide a default API so that external users don't have to specify it
+    const value = useMemo<AppContext>(() => ({ api: getDefaultAPIWrapper(), ...initialContext }), [initialContext]);
```

Comment:
we've had a few cases where the actions for samples or picklists needed to be exported from the ui-components package so that they could be used on the LK module side instead of being redefined there. Does this mean that if the LK app used this context provider it could just use the action via the api wrapper here instead of having to export each action separately in index.ts?

Reply:
Good question. In quite a few cases our "actions" are the same thing as what we call "API"s. It will likely be a more case-by-case basis kind of thing but we should discuss the merits of having explicitly exported methods versus context-provided methods (some are more obvious than others) and how to choose.

Reply:
Yes, you would grab them via:

```typescript
const appContext = useAppContext();
const samplesApi = appContext.api.samples;
const [sampleStatuses, setSampleStatuses] = useState();

useEffect(() => {
    samplesApi.getSampleStatuses().then(statuses => setSampleStatuses(statuses));
}, [samplesApi]);
```

Reply:
If you find yourself needing the samples API often you can write a helper:

```typescript
function useSamplesAPI() {
    const appContext = useAppContext();
    return appContext.api.samples;
}
```

Reply:
+1